packages/dds/tree/docs/wip/Schema.ts

```diff
@@ -0,0 +1,621 @@
+/**
+ * Example internal schema representation types.
+ */
+
+/**
+ * Stable identifier, used when persisting data.
+ *
+ * This intentionally does not support any concept of versioning or name-spacing:
+ * users of it can include versions and namespaces in if they want, or features for them could be added later.
+ *
+ * Can be used either:
+ * 1. Record just this in the persisted data.
+ * When loading the data, see if the loader has a schema with a matching name, and if so, use that.
+ * Optionally validate loaded data against schema.
+ * 2. Persist the whole schema. Use the identifier to associate it with schema when loading to check that the schema match.
+ */
+export type SchemaIdentifier = string & {
+    readonly SchemaIdentifier: "3965006b-66d0-41d6-bb04-bd6873b528b4";
+};
+
+/**
+ * SchemaIdentifier for a Tree.
+ * Also known as "Definition"
+ */
+export type TreeSchemaIdentifier = SchemaIdentifier & {
+    readonly TreeSchemaIdentifier: "ffc4b4b6-a4d8-4479-9636-fe6c6a1a4a7f";
+};
+
+/**
+ * Key / Name / Label for a field which is scoped to a specific TreeSchema.
+ */
+export type LocalFieldKey = string & {
+    readonly TreeSchemaIdentifier: "1108736b-ebb5-4924-9a25-d4dbabd83fc5";
+};
+
+/**
+ * SchemaIdentifier for a Field "global field",
+ * meaning a field which has the same meaning for all usages withing the document (not scoped to a specific TreeSchema like LocalFieldKey).
+ *
+ * Note that the implementations should ensure that GlobalFieldKeys can never collide with LocalFieldKeys.
+ * This can either be done in several ways
+ * (keeping the two classes of fields separate, namespaceing/escaping, compressing one into numbers and leaving the other strings, etc.)
+ */
+export type GlobalFieldKey = SchemaIdentifier & {
+    readonly TreeSchemaIdentifier: "7cc802f9-3360-4176-bbbf-cab23320d80b";
+};
+
+/**
+ * Schema for how many children are allowed in a field.
+ */
+export enum Multiplicity {
+    /**
+     * Exactly one item.
+     */
+    Value,
+    /**
+     * 0 or 1 items.
+     */
+    Optional,
+    /**
+     * 0 or more items.
+     */
+    Sequence,
+    /**
+     * Exactly 0 items.
+     */
+    Forbidden,
```

Comment:
Do you have a use case in mind for 'Forbidden'?

The only thing that comes to mind for me is to prevent mistakes in a discriminated union (e.g., I distinguish between types 'X' and 'Y' by knowing that field 'foo' only exists on X.  For safety, I could explicitly forbid 'foo' on Y.)